public/app/plugins/datasource/cloud-monitoring/components/QueryEditor.tsx

```diff
@@ -23,20 +25,29 @@ export const QueryEditor = (props: Props) => {
   const [modalIsOpen, setModalIsOpen] = useState<boolean>(false);
   const [migratedQuery, setMigratedQuery] = useState<CloudMonitoringQuery | undefined>();
   const query = useMemo(() => {
+    // We should only set the migrated query on the first load of the editor. On subsequent re-renders there's no need to migrate again.
     if (!migratedQuery) {
-      const migratedQuery = datasource.migrateQuery(oldQ);
-      setMigratedQuery(migratedQuery);
+      const migrated = datasource.migrateQuery(oldQ);
       // Update the query once the migrations have been completed.
-      onChange({ ...migratedQuery });
-      return migratedQuery;
+      setMigratedQuery(migrated);
+      return migrated;
     }
 
-    if (migratedQuery) {
+    // If the migrated query is the same as the updated query we can return the migrated query
+    // Otherwise the query has changed so we return the updated query
+    if (migratedQuery && deepEqual(migratedQuery, oldQ)) {
```

Comment:
Don't we need to account for `oldQ` changing? The relevant conditions would be:

1. it's a re-render and our query has changed
2. `migratedQuery` is set (to the old query), so we don't run the migration specified in the block.
3. This block doesn't run since the `deepEqual` fails
4. We return the unmigrated query.

I think  the last line should run the migration and return that result?

This hook would look like this then to avoid re-running the migration every time:

```ts
 if (migratedQuery && deepEqual(migratedQuery, oldQ)) {
      return migratedQuery;
    }
    const migrated = datasource.migrateQuery(oldQ);
    setMigratedQuery(migrated);
    return migrated;
```

Reply:
Excellent feedback @asimpson!